File: sources/go/stack/LinkedStack.go
/*
 *  链式栈
 */

package stack

import (
	"fmt"
)

// 链栈节点
type LinkedNode struct {
	data 		interface{}
	next		*LinkedNode
}

// 链栈结构体
type LinkedStack struct {
	top 		*LinkedNode				// 栈顶指针
	length		int						// 栈的长度
}

func NewLinkedStack() *LinkedStack{
	return &LinkedStack{
		top:	nil,
		length:	0,
	}
}

// 压栈操作
func (s *LinkedStack)Push(e interface{}) {
	node := &LinkedNode{
		data: e,
		next: s.top,
	}
	s.top = node
	s.length++
}

// 弹栈操作：返回出栈的值
func (s *LinkedStack)Pop() interface{}{

	if s.top == nil  {
		fmt.Println("空栈无法pop")
		return nil
	}

	e := s.top.data
	node := s.top
	s.top = node.next
	s.length--
	return e

}

// 栈顶元素
func (s *LinkedStack)Top() interface{}{
	if s.top == nil  {
		fmt.Println("空栈无栈顶元素")
		return nil
	}
	return s.top
}

File: sources/js/stack/linkedStack.js
/*
 * @Name: 栈
 * @Description: 链表实现栈
 * @Author: yaya
 * @Date: 2020-01-15 11:36:26
 * @Editors: yaya
 */

class Node{
    constructor(data){
        this.data = data;
        this.next = null;
    }
}

class LinkedStack{
    constructor(){
        this.node = new Node(null);
        this.top = this.node;
        this.length = 0;
    }
    
    // 压栈
    push(el){
        let newNode = new Node(el);

        let currentNode = this.node;

        newNode.next = currentNode;
        this.node = newNode;
        this.top = newNode;
        this.length++;
    }

    // 出栈弹栈
    pop(){
        if(this.length == 0){
            return null;
        }

        // 取头部
        let firstNode  = this.node;

        if(this.length == 1){
            this.node = null;
            this.top = null;
            this.length--;
            return firstNode.data;
        }

        let secondNode = this.node.next;

        this.node = secondNode;
        this.top = secondNode;
        this.length --;

        return firstNode.data;

    }

    // 获取栈顶元素
    topEl(){
        if(this.length == 0){
            console.log('栈为空栈');
            return ;
        }
        
        return this.top.data; 
    }
}

module.exports = LinkedStack